### ISSUE TITLE
Infinite elments revitalisation

### ISSUE THREAD

**Issue by BalticPinguin** [libMesh/libmesh#830]
The first part I already mentioned some time ago via email: If you run the miscellaneous ex1 in debug-mode, you get the following message:
``` 
 Building Infinite Elements:
  updating element neighbor tables...
  collecting boundary sides...
  found 96 inner and 0 outer boundary faces
  added 0 infinite elements and 0 nodes to the mesh 
```
So in fact, there are no infinite elements. To fix this, I would suggest the following patch:
[infinite.patch.txt](https://github.com/libMesh/libmesh/files/136317/infinite.patch.txt)

With this, two problems occur: First, exodus refuses to work here; Roy Stogner explained me that " we use one block per subdomain id, and Exodus doesn't allow more than one element type
per block." The easiest way probably would be to remove the statement or to use an other format at this point; maybe in a longer term one could change the format such that exodus is possible here as well!?

However, even after this the program fails, with the error code
```/usr/include/c++/4.8/debug/vector:353:error: attempt to subscript container 
    with out-of-bounds index 0, but container only holds 0 elements.

Objects involved in the operation:
sequence "this" @ 0x0x1475d30 {
  type = NSt7__debug6vectorINS0_IdSaIdEEESaIS2_EEE;
}
Aborted (core dumped)```
which is because the elements Ss and St in combine_base_radial() (src/fe/inf_fe.C) are empty:
`(gdb) p Ss.size()
$1 = 0
(gdb) p St.size()
$2 = 0
(gdb) p mode.size()
$3 = 4`
As a fix, I would suggest the following:
[infinite2.patch.txt](https://github.com/libMesh/libmesh/files/136339/infinite2.patch.txt)

which made the example working for me finally.
Best,
Hubert

**Comment by permcody** [comment]
On Thu, Feb 18, 2016 at 7:42 AM Hubert Weißmann <notifications@github.com>
wrote:

> The first part I already mentioned some time ago via email: If you run the
> miscellaneous ex1 in debug-mode, you get the following message:
>
>  Building Infinite Elements:
>   updating element neighbor tables...
>   collecting boundary sides...
>   found 96 inner and 0 outer boundary faces
>   added 0 infinite elements and 0 nodes to the mesh
>
> So in fact, there are no infinite elements. To fix this, I would suggest
> the following patch:
> infinite.patch.txt
> <https://github.com/libMesh/libmesh/files/136317/infinite.patch.txt>
>
> With this, two problems occur: First, exodus refuses to work here; Roy
> Stogner explained me that " we use one block per subdomain id, and Exodus
> doesn't allow more than one element type
> per block." The easiest way probably would be to remove the statement or
> to use an other format at this point; maybe in a longer term one could
> change the format such that exodus is possible here as well!?
>
How about simply creating another subdomain block if you have another
element type? That's the normal usage case for handling exodus format. It's
not an arbitrary restriction. It's so that exodus can reduce the
information it packs for each element. If you know the type for the whole
block, then you can simply list the connectivity id list for each element
and no other information is required.

**Comment by BalticPinguin** [comment]
yes, that sounds sensible. I have not dealt with the format so far but will have a look at it and try to write some code to patch it.

**Comment by permcody** [comment]
You shouldn't need to change anything about the writer or the format. Just
assign a different subdomain_id to each unique element type (at least) and
the writer will take care of the format.

**Comment by BalticPinguin** [comment]
It might be not that easy; I tried to fix it by the patch:
[misc1.patch.txt](https://github.com/libMesh/libmesh/files/142308/misc1.patch.txt)
but than, the error
```
Unsupported element type: 22
Stack frames: 8
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: libMesh::ExodusII_IO_Helper::ElementMaps::assign_conversion(libMesh::ElemType)
3: libMesh::ExodusII_IO_Helper::write_elements(libMesh::MeshBase const&, bool)
4: libMesh::ExodusII_IO::write(std::string const&)
5: ./example-dbg() [0x430883]
6: __libc_start_main
7: ./example-dbg() [0x4302a9]
[0] src/mesh/exodusII_io_helper.C, line 2295, compiled Feb 18 2016 at 17:01:56
```
appears.
Indeed, printing the value of a variable is not that simple for infinite elements, since it is not just the element shape function but the esf multiplied by an exponential with phase and a weighting function.
Maybe therefore no one implemented the respective routines so far?

**Comment by jwpeterson** [comment]
In order to not lose the work that you've done so far, I pushed a [branch](https://github.com/jwpeterson/libmesh/commits/infinite_elements) with your commits.  Since git could not understand the format of your first two patches, I had to recreate them by hand.  Please pull this branch, check that the changes are correct, and make more commits in your libmesh fork if you need to continue the work.

As far as the issue with actually writing infinite elements to an Exodus file, I'm pretty sure we don't want to do that, as there is no way of viewing them in Paraview or other visualization software that I know of.  Instead, the Exodus reader probably just needs to be patched so that it ignores them.  I can take a crack at this, but probably not until this weekend some time, so if you get to it first that would be great.

**Comment by jwpeterson** [comment]
Your infinite elements fixes were merged into master in 9c268ca.  I'll close this for now, but feel free to re-open and/or start another issue if you find other infinite elements issues.
